docs/proposals/01-extensibility.md

```diff
@@ -0,0 +1,879 @@
+
+## Summary
+
+Gardener has evolved to a large compound of packages containing lots of highly specific knowledge which makes it very hard to extend (supporting a new cloud provider, new OS, ..., or behave differently depending on the underlying infrastructure).
+
+This proposal aims to move out the cloud-specific implementations (called "(cloud) botanists") and the OS-specifics into dedicated controllers, and simultaneously to allow deviation from the standard Gardener deployment.
+
+## Motivation
+
+Currently, it is too hard to support additional cloud providers or operation systems/distributions as everything must be done in-tree which might affect the implementation of other cloud providers as well.
+The various conditions and branches make the code hard to maintain and hard to test.
+Every change must be done centrally, requires to completely rebuild Gardener, and cannot be deployed individually. Similar to the motivation for Kubernetes to extract their cloud-specifics into dedicated cloud-controller-managers or to extract the container/storage/network/... specifics into CRI/CSI/CNI/..., we aim to do the same right now.
+
+### Goals
+
+* Gardener does not contain any cloud-specific knowledge anymore but defines a clear contract allowing external controllers (botanists) to support different environments (AWS, Azure, GCP, ...).
+* Gardener does not contain any operation system-specific knowledge anymore but defines a clear contract allowing external controllers to support different operation systems/distributions (CoreOS, SLES, Ubuntu, ...).
+* It shall become much easier to move control planes of Shoot clusters between Seed clusters ([#232](https://github.com/gardener/gardener/issues/232)) which is a necessary requirement of an automated setup for the Gardener Ring ([#233](https://github.com/gardener/gardener/issues/233)).
+
+### Non-Goals
+
+* We want to also factor out the specific knowledge of the addon deployments (nginx-ingress, kubernetes-dashboard, ...), but we already have dedicated projects/issues for that: https://github.com/gardener/bouquet and [#246](https://github.com/gardener/gardener/issues/246). We will keep the addons in-tree as part of this proposal and tackle their extraction separately.
+* We do not want to make the Gardener a plain workflow engine that just executes a given template (which indeed would allow to be generic, open, and extensible in their highest forms but which would end-up in building a "programming/scripting language" inside a serialization format (YAML/JSON/...)). Rather, we want to have well-defined contracts and APIs, keeping Gardener responsible for the clusters management.
+
+## Proposal
+
+Gardener heavily relies on and implements Kubernetes principles, and its ultimate strategy is to use Kubernetes wherever applicable.
+The extension concept in Kubernetes is based on (next to others) `CustomResourceDefinition`s as well as `ValidatingWebhookConfiguration`s and `MutatingWebhookConfiguration`s.
```

Comment:
I like your idea mentioned in the third approach: Initializers instead of the hooks. Will think about that.

Reply:
I like that approach as well, but how would that work in terms of pending initialisers? Would Gardener know from registration (complex) or be based on `initializerConfiguration`, but ten many initialisers may be involved not doing anything useful (e.g. backup & restore wants only to plugin into the etcd statefulset.

Reply:
Gardener wouldn't know them, it would be based on `InitializerConfiguration` which must be setup by the controllers requiring to initialize the objects. Without having looked further I'd say that we can motivate both (webhooks and initializers) and recommend initializers for long-running operations while webhooks are fine for small modifications on a resource. I still need to play it through with the current flow to give a better feedback to the idea.